plugins/fuse/server/server_fuse.py

```diff
@@ -0,0 +1,492 @@
+            attr['st_size'] = 0
+        else:
+            resource = self._getPath(path)
+            attr = self._stat(resource['document'], resource['model'])
+        if attr.get('st_blksize') and attr.get('st_size'):
+            attr['st_blocks'] = int(
+                (attr['st_size'] + attr['st_blksize'] - 1) / attr['st_blksize'])
+        return attr
+
+    def read(self, path, size, offset, fh):
+        """
+        Read a block of bytes from a resource.
+
+        :param path: path within the fuse.  Ignored, since the fh parameter
+            must be valid.
+        :param size: maximum number of bytes to read.  There may be less if
+            this is near the end of the file.
+        :param offset: the offset within the file to read.
+        :param fh: an open file handle.
+        :returns: a block of up to <size> bytes.
+        """
+        with self.openFilesLock:
+            if fh not in self.openFiles:
+                raise fuse.FuseOSError(errno.EBADF)
+            info = self.openFiles[fh]
+        with info['lock']:
+            handle = info['handle']
+            handle.seek(offset)
+            return handle.read(size)
+
+    def readdir(self, path, fh):
+        """
+        Get a list of names within a directory.
+
+        :param path: path within the fuse.
+        :param fh: an open file handle.  Ignored, since path is always
+            specified.
+        :returns: a list of names.  This always includes . and ..
+        """
+        path = path.rstrip('/')
+        result = [u'.', u'..']
+        if path == '':
+            result.extend([u'collection', u'user'])
+        elif path in ('/user', '/collection'):
+            model = path[1:]
+            if self.force:
+                docList = self.model(model).find({}, sort=None)
+            else:
+                docList = self.model(model).list(user=self.user)
+            for doc in docList:
+                result.append(self._name(doc, model))
+        else:
+            resource = self._getPath(path)
+            result.extend(self._list(resource['document'], resource['model']))
+        return result
+
+    def open(self, path, flags):
+        """
+        Open a path and return a descriptor.
+
+        :param path: path within the fuse.
+        :param flags: a combination of O_* flags.  This will fail if it is not
+            read only.
+        :returns: a file descriptor.
+        """
+        resource = self._getPath(path)
+        if resource['model'] != 'file':
+            return super(ServerFuse, self).open(path, flags)
+        if flags & (os.O_APPEND | os.O_ASYNC | os.O_CREAT | os.O_DIRECTORY |
+                    os.O_EXCL | os.O_RDWR | os.O_TRUNC | os.O_WRONLY):
+            raise fuse.FuseOSError(errno.EROFS)
+        info = {
+            'path': path,
+            'handle': File().open(resource['document']),
+            'lock': threading.Lock(),
+        }
+        with self.openFilesLock:
+            fh = self.nextFH
```

Comment:
Sorry,  i don't think i'm following the need for the nextFH file indexing,  why not just use info['handle'].fileno()?

Reply:
The handle may not refer to a physical file -- it could be an S3 assetstore file, or a database file, etc.  It may not have its own fileno.